src/acvp.c

```diff
@@ -803,6 +807,334 @@ ACVP_RESULT acvp_load_kat_filename(ACVP_CTX *ctx, const char *kat_filename) {
+    reg_array = json_value_get_array(val);
+    obj = json_array_get_object(reg_array, n);
+    if (!obj) {
+        ACVP_LOG_ERR("JSON obj parse error");
+        goto end;
+    }
+
+    /*
+     * This is the identifiers provided by the server
+     * for this specific test session!
+     */
+    test_session_url = json_object_get_string(obj, "url");
+    if (test_session_url) {
+        ctx->session_url = calloc(ACVP_ATTR_URL_MAX + 1, sizeof(char));
+        strcpy_s(ctx->session_url, ACVP_ATTR_URL_MAX + 1, test_session_url);
+    } else {
+        ACVP_LOG_WARN("Missing session URL, results will not be POSTed to server");
+        goto end;
+    }
+
+    jwt = json_object_get_string(obj, "jwt");
+    if (jwt) {
+        ctx->jwt_token = calloc(ACVP_JWT_TOKEN_MAX + 1, sizeof(char));
+        strcpy_s(ctx->jwt_token, ACVP_JWT_TOKEN_MAX + 1, jwt);
+    } else {
+        ACVP_LOG_WARN("Missing JWT, results will not be POSTed to server");
+        goto end;
+    }
+
+    vect_sets = json_object_get_array(obj, "vectorSetUrls");
+    vs_cnt = json_array_get_count(vect_sets);
+    for (i = 0; i < vs_cnt; i++) {
+        char *vsid_url = (char*)json_array_get_string(vect_sets, i);
+
+        if (!vsid_url) {
+            ACVP_LOG_WARN("No vsId URL, results will not be POSTed to server");
+            goto end;
+        }
+
+        rv = acvp_append_vsid_url(ctx, vsid_url);
+        if (rv != ACVP_SUCCESS) goto end;
+        ACVP_LOG_INFO("Received vsid_url=%s", vsid_url);
+    }
+
+    n++;        /* bump past the version or url, jwt, url sets */
+    obj = json_array_get_object(reg_array, n);
+    if (!obj) {
+        ACVP_LOG_ERR("JSON obj parse error");
+        goto end;
+    }
+
+    vs_entry = ctx->vsid_url_list;
+    if (!vs_entry) {
+        goto end;
+    }
+
+    while (obj) {
+        if (!vs_entry) {
+            goto end;
+        }
+        /* Process the kat vector(s) */
+        rv  = acvp_dispatch_vector_set(ctx, obj);
+        if (rv != ACVP_SUCCESS) {
+            ACVP_LOG_ERR("KAT dispatch error");
+            json_value_free(val);
+            goto end;
+        }
+        ACVP_LOG_STATUS("Write vector set response vsId: %d", ctx->vs_id);
+
+        /* 
+         * Convert the JSON from a fully qualified to a value that can be 
+         * added to the file. Kind of klumsy, but it works.
+         */
+        kat_array = json_value_get_array(ctx->kat_resp);
+        kat_val = json_array_get_value(kat_array, 1);
+        if (!kat_val) {
+            ACVP_LOG_ERR("JSON val parse error");
+            goto end;
+        }
+        json_result = json_serialize_to_string_pretty(kat_val, NULL);
+        file_val = json_parse_string(json_result);
+        json_free_serialized_string(json_result);
+
+        /* track first vector set with file count */
+        if (n == 1) {
+            ACVP_STRING_LIST *vs_entry = NULL;
+
+            rsp_val = json_value_init_object();
+            rsp_obj = json_value_get_object(rsp_val);
```

Comment:
Can you not use the val/obj that you parsed earlier from the req file? Instead of making new ones here

Reply:
I am trying to be very explicit there since there are several values used in this function